tensorboard/data/server/cli/dynamic_logdir.rs

```diff
@@ -52,29 +67,43 @@ impl DynLogdir {
     /// [seanmonstar/reqwest#1017].
     ///
     /// [seanmonstar/reqwest#1017]: https://github.com/seanmonstar/reqwest/issues/1017
-    pub fn new(path: PathBuf) -> Option<Self> {
+    pub fn new(path: PathBuf) -> Result<Self, Error> {
         let path_str = path.to_string_lossy();
-        let gcs_path = match path_str.strip_prefix("gs://") {
-            // Assume that anything not starting with `gs://` is a path on disk.
-            None => return Some(DynLogdir::Disk(DiskLogdir::new(path))),
-            Some(p) => p,
-        };
+        if path_str.is_empty() {
+            return Err(Error::EmptyLogdir);
+        }
+
+        let protocol_parts: Vec<&str> = path_str.splitn(2, "://").collect();
+        if protocol_parts.len() < 2 || !is_protocol(protocol_parts[0]) {
+            // Interpret as path on disk.
+            return Ok(DynLogdir::Disk(DiskLogdir::new(path)));
+        }
+
+        let (protocol, subpath) = (protocol_parts[0], protocol_parts[1]);
+        match protocol {
+            "gs" => Self::new_gcs(subpath),
+            _ => Err(Error::UnknownProtocol {
+                protocol: protocol.to_string(),
+                full_logdir: path,
+            }),
+        }
+    }
+
+    fn new_gcs(gcs_path: &str) -> Result<Self, Error> {
         let mut parts = gcs_path.splitn(2, '/');
         let bucket = parts.next().unwrap().to_string(); // splitn always yields at least one element
         let prefix = parts.next().unwrap_or("").to_string();
-        let creds = gcs::Credentials::from_disk().unwrap_or_else(|e| {
-            warn!("Using anonymous GCS credentials: {}", e);
-            Default::default()
-        });
-        let client = match gcs::Client::new(creds) {
-            Err(e) => {
-                error!("Could not open GCS connection: {}", e);
-                return None;
-            }
-            Ok(c) => c,
-        };
-        Some(DynLogdir::Gcs(gcs::Logdir::new(client, bucket, prefix)))
+        let client = gcs::Client::new(gcs::Credentials::from_disk()?)?;
+        Ok(DynLogdir::Gcs(gcs::Logdir::new(client, bucket, prefix)))
+    }
+}
+
+fn is_protocol(s: &str) -> bool {
```

Comment:
Link to https://tools.ietf.org/html/rfc3986#section-3.1 if that's what we're implementing here?

FWIW, the logic in TB proper (for no-TF mode; haven't checked real TF) doesn't validate the scheme/protocol; it only splits on "://".  Not a big deal but maybe worth trying to harmonize these two since it's a little odd to have the behavior differ.

Reply:
Yeah, 3986 is meant to update 1738 so I think it's the better reference.  SGTM to do it the correct way especially since TF also does that, and we can fix the no-TF one at some point.

Reply:
Ah, got it. I see the “Updates” tag now. I’ve always had trouble
navigating the maze of URL RFCs, so this will help; thanks.